app/ide-desktop/lib/dashboard/src/authentication/src/authentication/providers/session.tsx

```diff
@@ -48,10 +83,47 @@ export function SessionProvider(props: SessionProviderProps) {
       setInitialized(true);
       return innerSession;
     },
-    [userSession]
+    [refresh, userSession]
   );
 
+  /** Register an effect that will listen for authentication events. When the event occurs, we
+   * will refresh or clear the user's session, forcing a re-render of the page with the new
+   * session.
+   *
+   * For example, if a user clicks the signout button, this will clear the user's session, which
+   * means we want the login screen to render (which is a child of this provider). */
+  react.useEffect(() => {
+    const listener: listen.ListenerCallback = (event) => {
+      switch (event) {
+        case listen.AuthEvent.signIn: {
+          refreshSession();
+          break;
+        }
+        case listen.AuthEvent.customOAuthState:
+        case listen.AuthEvent.cognitoHostedUi: {
+          /** AWS Amplify doesn't provide a way to set the redirect URL for the OAuth flow, so
+           * we have to hack it by replacing the URL in the browser's history. This is done
+           * because otherwise the user will be redirected to a URL like `enso://auth`, which
+           * will not work.
+           *
+           * See:
+           * https://github.com/aws-amplify/amplify-js/issues/3391#issuecomment-756473970 */
+          window.history.replaceState({}, "", mainPageUrl);
```

Comment:
you can make the docs smaller by moving "see" to previous line as `will not work, see: ....`. Minor thing though.